### DISCUSSION TITLE
Stuart Feedback

### DISCUSSION THREAD

**Discussion by makubacki** [tianocore/edk2#4369] [General]
A tool called Stuart is integrated with edk2 to help provide a consistent build and dependency management interface across different user environments.

Quick references:

- Repo that hosts Stuart - [edk2-pytool-extensions](https://github.com/tianocore/edk2-pytool-extensions)
- User documentation area - [User Documentation](https://www.tianocore.org/edk2-pytool-extensions/)
- TianoCore Specific Stuart Build Instructions - [How to Build with Stuart](https://github.com/tianocore/tianocore.github.io/wiki/How-to-Build-With-Stuart)

Feedback has been received on numerous occasions that Stuart:

1. Does not integrate well into some developer workflows - particularly on Linux
2. Has dependencies (like .NET) that are complicated to install - especially on various Linux distros
3. Has a CLI that is difficult to understand/use
4. Has a configuration process that is difficult to understand/use
5. Has a poor implementation
6. Is redundant/unnecessary

and so on...

This discussion was created to gather constructive feedback that can be channeled into actionable improvements. The tool aspires to provide a Python-based build environment that can accommodate common developer flows in the edk2 project, and this feedback will hopefully help bring it closer to achieving that.

---

I understand it can be frustrating to use something that may have fallen short of expectations. However, please try to keep feedback constructive, concise, and respectful.

**Comment by ghost** [comment]
One issue for me is having to pass the toolchain using `TOOL_CHAIN_TAG=` instead of `-t` (related: -t appears to have been redefined to mean "target" instead of "toolchain").
Also, specifying the build type using `TARGET={RELEASE,NOOPT,DEBUG}` doesn't appear to be documented.

Errors aren't always propagated:
```
(venv) bcran@cube:~/src/uefi/edk2$ stuart_update -c .pytool/CISettings.py
...
SECTION - 	Updated/Verified 3 dependencies
SECTION - Second pass update of environment
ERROR - Could not find appropriate folder for mu-uncrustify-release. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for mu_nasm. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for edk2-acpica-iasl. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for mu-uncrustify-release. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for mu_nasm. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for edk2-acpica-iasl. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for mu-uncrustify-release. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for mu_nasm. Host(os='Linux', arch='x86', bit='64')
ERROR - Could not find appropriate folder for edk2-acpica-iasl. Host(os='Linux', arch='x86', bit='64')
Updating. Done
SECTION - 	Updated/Verified 3 dependencies
SECTION - Summary
PROGRESS - Success
```
Also, when using Edk2ToolsBuild.py, if I remove the brotli submodule code:
```
(venv) bcran@cube:~/src/uefi/tmp/edk2$ python3 BaseTools/Edk2ToolsBuild.py -t GCC5
SECTION - Init Self Describing Environment
SECTION - Loading Plugins
SECTION - Start Invocable Tool
Traceback (most recent call last):
  File "/home/bcran/src/uefi/tmp/edk2/BaseTools/Edk2ToolsBuild.py", line 177, in <module>
    main()
  File "/home/bcran/src/uefi/tmp/edk2/BaseTools/Edk2ToolsBuild.py", line 173, in main
    Edk2ToolsBuild().Invoke()
  File "/home/bcran/src/uefi/tmp/edk2/venv/lib/python3.11/site-packages/edk2toolext/base_abstract_invocable.py", line 236, in Invoke
    retcode = self.Go()
              ^^^^^^^^^
  File "/home/bcran/src/uefi/tmp/edk2/BaseTools/Edk2ToolsBuild.py", line 148, in Go
    raise Exception("Failed to build.")
Exception: Failed to build.
```
Looking in BaseTools/BaseToolsBuild/BASETOOLS_BUILD.txt there are no ERROR lines, even at:
```
INFO - BrotliCompress.c:20:10: fatal error: ./brotli/c/common/constants.h: No such file or directory
INFO -    20 | #include "./brotli/c/common/constants.h"
INFO -       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
INFO - compilation terminated.
```

**Reply by makubacki** [reply]
Thank you for taking time to collect this info.

I was also thrown off by `-t` not corresponding to the tool chain.

edk2 is [currently using](https://github.com/tianocore/edk2/blob/master/pip-requirements.txt) edk2-pytool-extensions 0.21.8. I noticed a similar issue with GCC compilation errors not being picked up and made a change in [0.21.9](https://github.com/tianocore/edk2-pytool-extensions/releases/tag/v0.21.9) to improve that. Do you have time to see if that helps with the issue you encountered?

**Reply by ghost** [reply]
> edk2 is [currently using](https://github.com/tianocore/edk2/blob/master/pip-requirements.txt) edk2-pytool-extensions 0.21.8. I noticed a similar issue with GCC compilation errors not being picked up and made a change in [0.21.9](https://github.com/tianocore/edk2-pytool-extensions/releases/tag/v0.21.9) to improve that. Do you have time to see if that helps with the issue you encountered?

It doesn't appear to help with BaseTools build errors.
```
(venv) bcran@cube:~/src/uefi/edk2$ pip3 list | grep edk2
edk2-basetools         0.1.45
edk2-pytool-extensions 0.23.2
edk2-pytool-library    0.14.1
```
```
INFO - make[2]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C/VfrCompile'
INFO - make[2]: *** No rule to make target 'brotli/c/common/constants.h', needed by 'BrotliCompress.o'.  Stop.
INFO - make[2]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C/BrotliCompress'
INFO - make[1]: *** [GNUmakefile:79: BrotliCompress] Error 2
INFO - make[1]: *** Waiting for unfinished jobs....
INFO - make[2]: Nothing to be done for 'all'.
INFO - make[2]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C/TianoCompress'
INFO - make[2]: Nothing to be done for 'all'.
INFO - make[2]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C/VolInfo'
INFO - make[2]: Nothing to be done for 'all'.
INFO - make[2]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C/LzmaCompress'
INFO - make[2]: Nothing to be done for 'all'.
INFO - make[2]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C/DevicePath'
INFO - make[1]: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools/Source/C'
INFO - make: *** [GNUmakefile:19: Source/C] Error 2
INFO - make: Leaving directory '/home/bcran/src/uefi/edk2/BaseTools'
INFO - ------------------------------------------------
INFO - --------------Cmd Output Finished---------------
INFO - --------- Running Time (mm:ss): 00:00 ----------
INFO - ----------- Return Code: 0x00000002 ------------
INFO - ------------------------------------------------
```

**Reply by makubacki** [reply]
Thanks for checking. An issue here is that the GNUmakefile errors need to be recognized and logged as `ERROR`. I filed a bug to track it - https://github.com/tianocore/edk2-pytool-extensions/issues/539.

**Reply by makubacki** [reply]
> One issue for me is having to pass the toolchain using `TOOL_CHAIN_TAG=` instead of `-t` (related: -t appears to have been redefined to mean "target" instead of "toolchain"). Also, specifying the build type using `TARGET={RELEASE,NOOPT,DEBUG}` doesn't appear to be documented.

You can eventually see some help:

```
(venv3.10) ➜  edk2 git:(master) ✗ stuart_ci_build -c .pytool/CISettings.py --help
usage: stuart_ci_build [-h] [-p PACKAGELIST] [-a REQUESTED_ARCH] [-t REQUESTED_TARGET] [-force_piptools | -no_piptools]
                       [--build-config BUILD_CONFIG] [--verbose]

options:
  -h, --help            show this help message and exit
  -p PACKAGELIST, --pkg PACKAGELIST, --pkg-dir PACKAGELIST
                        Optional - A package or folder you want to update (workspace relative).Can list multiple by doing -p
                        <pkg1>,<pkg2> or -p <pkg3> -p <pkg4>
  -a REQUESTED_ARCH, --arch REQUESTED_ARCH
                        Optional - CSV of architecutres requested to update. Example: -a X64,AARCH64
  -t REQUESTED_TARGET, --target REQUESTED_TARGET
                        Optional - CSV of targets requested to update. Example: -t DEBUG,NOOPT
  -force_piptools, --fpt
                        Force the system to use pip tools
  -no_piptools, --npt   Force the system to not use pip tools
  --build-config BUILD_CONFIG
                        Provide shell variables in a file
  --verbose, --VERBOSE, -v
                        verbose

CLI Env Guide:
  <key>=<value>              - Set an env variable for the pre/post build process
  <key>                      - Set a non-valued env variable for the pre/post build process
  BLD_*_<key>=<value>        - Set a build flag for all build types
                               (key=value will get passed to build process)
  BLD_*_<key>                - Set a non-valued build flag for all build types
  BLD_<TARGET>_<key>=<value> - Set a build flag for build type of <target>
                               (key=value will get passed to build process for given build type)
  BLD_<TARGET>_<key>         - Set a non-valued build flag for a build type of <target>
```

That also works for other commands supported by `.pytool/CISettings.py` like `stuart_ci_setup`.

Note that I had to know to use `stuart_ci_build` instead of `stuart_build` (because this is a "CI" file - `.pytool/CISettings.py`). If I had tried to use `stuart_build`:

```
(venv3.10) ➜  edk2 git:(master) ✗ stuart_build -c .pytool/CISettings.py --help 
Unable to use .pytool/CISettings.py as a BuildSettingsManager
Did you mean to use a different kind of invocable?
Traceback (most recent call last):
  File "/home/mikuback/src/edk2/venv3.10/lib/python3.10/site-packages/edk2toolext/edk2_invocable.py", line 360, in ParseCommandLineOptions
    self.PlatformSettings = locate_class_in_module(
TypeError: 'NoneType' object is not callable

...<Python traceback continues on for a while>
```

The text "Did you mean to use a different kind of invocable?" is a hint but probably not useful to someone looking for help.

Did finding this information contribute to the issue you were having? Is the information helpful?

**Reply by makubacki** [reply]
Issue filed for `-t` confusion: https://github.com/tianocore/edk2-pytool-extensions/issues/540

**Reply by Javagedes** [reply]
https://github.com/tianocore/edk2-pytool-extensions/issues/540 has been resolved. Additional error logging has been added when either TOOL_CHAIN_TAG or TARGET has not been set, along with a graceful exit. This will be available when v0.23.3 is released.

**Reply by ghost** [reply]
Thanks!

**Comment by ghost** [comment]
@makubacki I was wondering if you could take a look at my stuart_pr_eval failure at https://github.com/tianocore/edk2-pytool-extensions/discussions/550 ? I don't know why it's not working.

**Reply by makubacki** [reply]
I replied. Feel free to follow up there if more help is needed.

**Comment by osteffenrh** [comment]
> * Does not integrate well into some developer workflows - particularly on Linux

A wish of mine would be an "offline" mode, where it does not download the toolchain (gcc, nasm, etc) but uses the tools it can find in `PATH`. That is useful for Linux distribution specific work where we want to use the tools of the distro.

I started looking into this a while back but then got distracted.

> * Has dependencies (like .NET) that are complicated to install - especially on various Linux distros

I get that .NET is convenient in come cases and it is easily available on most Linux environments, but it is a large dependency compared to what it is used for in the end (downloading dependencies via nuget, I think? Or does it do other things too?). It adds quite some size to the container images that we use in the (Linux) CI.

**Reply by makubacki** [reply]
Offline mode seems reasonable, I've heard it before. We're also reducing .NET usage. @Javagedes will follow up.

**Reply by spbrogan** [reply]
@osteffenrh - the online/offline and even .NET dependency comes down to how a platform uses stuart.  For example, if you don't leverage any nuget package based dependencies then you don't actually need .NET or nuget.  As @makubacki mentioned because this has been a pain point for some edk2 developers we are trying to move edk2 away from nuget based dependencies.  

The platform controls the dependencies and tools it uses.  You can control this through the scopes.  You could easily create an option for an edk2 platform (like ovmf, armvirt, emulatorpkg) or your own platform that disables the scopes and makes it work with tools on the path.  I don't believe this is a scalable or sustainable practice in that an end user has no idea what versions of tools are supported and thus we didn't do that when setting up the CI platforms in Edk2.  One of the best value points for pytools is that it lets you define a repeatable and reproducible environment using artifacts that are checked into the source tree.   But regardless, pytools shouldn't need any changes to support offline mode.  Again, I am not a fan of this but can understand it could be ideal or useful for some projects and pytools in general should not require online usage if a "platform" doesn't require it.  Finally, since I am not sure we have built a platform using all external tools (in a long time) it is possible something has creeped in.  If so please let us know so we can fix it.

**Comment by Javagedes** [comment]
As requested from @mdkinney, We've added some stuart CLI usability improvements and will leave them open for a few days for anyone to leave comments.

Review each PR to see the changes, but overall they do the following:

- When using the `-h` flag, Print the available options for `-t` (target), `-p` (package), and `-a` (architecture) if a configuration file has been provided.
- Return a clean error if you provide an invalid command line argument (to include `-<flag>`, `--<flag>`, or `/<flag>`)
- Allow running a subset of tests with `stuart_ci_build` by setting `-d` or `--disable-all` and then turning back on the plugin you wish to run via the command line with `<Plugin>=run`

https://github.com/tianocore/edk2-pytool-extensions/pull/753
https://github.com/tianocore/edk2-pytool-extensions/pull/754
https://github.com/tianocore/edk2-pytool-extensions/pull/755

**Comment by mikebeaton** [comment]
@makubacki @Javagedes -

Was there any update on moving edk2 away from NuGet based dependencies?

I agree with @osteffenrh that it seems .NET (e.g. on Linux `mono-complete`, which is large) is used just to download NuGet dependencies, which in context are basically versioned zip files, making .NET a very heavyweight dependency to support this.

I don't know if there's been any thought to adding GitHub releases as a supported dependency type in https://github.com/tianocore/edk2-pytool-extensions/tree/master/edk2toolext/environment/extdeptypes ? These can be arbitrarily named releases, containing arbitrarily named zip files (not necessarily builds of source in the associated repo) which I believe is about the right amount of structure to conveniently replace the way nasm, iasl and so on are currently downloaded from NuGet. I've seen other projects use this as a place to store versions of dependencies.

**EDIT 1**: I guess GitHub release files can be fetched already, with the existing web dependency type (haven't yet tested, but it looks like it should be fine), but perhaps it still might be convenient to make explicit the support for named files within named releases in GitHub releases, with a new dependency type, rather then just folding this into a URL?

**EDIT 2**: A release can be from an arbitrarily named tag on a repo, I think it must contain the (auto-generated) zip of the repo source at that tag, but can have any other arbitrary added zip files - I'm suggesting naming the release in the former way, then adding the iasl, masm, etc. zip files in the latter way - probably from a repo without much source, perhaps just a README, for this purpose. This is a very rough-and-ready demo of the idea: https://github.com/mikebeaton/edk2_ci/releases/tag/edk2-release-test-v0.1

**Reply by Javagedes** [reply]
@mikebeaton Hello, as you mentioned, this can already be done with the web dependency exactly as requested as above, however it is not the most ergonomic. We have an example of this in mu_basecore, as we pre-compile our basetools rather than compile them locally:

https://github.com/microsoft/mu_basecore/blob/dev/202405/BaseTools/Bin/basetoolsbin_ext_dep.yaml
``` json
  "scope": "global",
  "id": "edk-tools-bin",
  "type": "web",
  "name": "Mu-Basetools",
  "source": "https://github.com/microsoft/mu_basecore/releases/download/v2023020002.0.3/basetools-v2023020002.0.3.tar.gz",
  "version": "v2023020002.0.3",
  "sha256": "7260efe6e4e734ac66b24d6dc5f4908bf7aa444a7827cb405a66131b7b75e86f",
  "internal_path": "/basetools/",
  "compression_type": "tar",
  "flags": ["set_shell_var", "set_path", "host_specific"],
  "var_name": "EDK_TOOLS_BIN"
```

I'd be happy to create an external dependency specific to github releases, what are you envisioning in terms of ergonomics compared to the web dependency?

I could see us converting the `source` to a `repository` and reducing it to  `microsoft/mu_basecore` (`tianocore/edk2` for edk2) and then building the final download path as:

``` python
path = f'https://github.com/{repository}/releases/download/{version}/{name}-{version}.tar.gz'
```

But I'd hope for a bit more ergonomic improvement between web dependencyies and github release dependencies if possible

Looking forward to your feedback. I can spend some time to get it up and running once we come up with something good.

**Reply by Javagedes** [reply]
The issue with this, however, is that it does not work well in our current use case, which is where we re-package a different tool with a version we wish to preserve, in a release that does not match that version.
e.g. the github release is `edk2-release-test-v0.1` but the zipped files are a different version (`20190215.0.0` and `2.15.5`). At that point I kind of feel like the web dependency would be better, as you specify the source path directly.

**Reply by mikebeaton** [reply]
Hi!

How about:

```
"type": "github-release",
"repo": "tianocore/edk2-ci",
"release": "mu_nasm_2.15.5",
"file": "mu_nasm_2.15.5.zip",
.
.
.
```

Which is then otherwise the same as a web dependency with `source`:

https://github.com/tianocore/edk2-ci/releases/download/mu_nasm_2.15.5/mu_nasm_2.15.5.zip

Equally:

```
"type": "github-release",
"repo": "tianocore/edk2-ci",
"release": "iasl.20190215.0.0",
"file": "iasl.20190215.0.0.zip",
.
.
.
```

Which acts like a web dependency with `source`:

https://github.com/tianocore/edk2-ci/releases/download/iasl.20190215.0.0/iasl.20190215.0.0.zip

Then the dependency format allows fetching an arbitrary file from an arbitrary release, but the proposed repo https://github.com/tianocore/edk2-ci keeps to one tag-plus-release, per version, per tool?

**Reply by Javagedes** [reply]
Yes that seems good to me. I'll prototype it.

**Comment by xpahos** [comment]
```
sudo apt-get install gcc-arm-linux-gnueabihf gcc-aarch64-linux-gnu lcov
curl -sL https://deb.nodesource.com/setup_18.x -o nodesource_setup.sh
sudo bash nodesource_setup.sh
node --dns-result-order=ipv4first \$(which npm) install -g cspell@latest
curl https://sh.rustup.rs -sSf \| sh
cargo install cargo-make
source ~/.bashrc
sudo apt install python3.11-venv
sudo apt install build-essential uuid-dev
sudo apt-get install libmono-system-net-http4.0-cil libmono-system-io-compression4.0-cil
python3.11 -m venv .venv
source .venv/bin/activate
git submodule update --init --recursive
pip3.11 install -r pip-requirements.txt --upgrade
stuart_setup -c .pytool/CISettings.py
stuart_update -c .pytool/CISettings.py
stuart_update -c OvmfPkg/PlatformCI/PlatformBuild.py
stuart_build -c OvmfPkg/PlatformCI/PlatformBuild.py -a IA32,X64 TOOL_CHAIN_TAG=GCC5
stuart_ci_build -c .pytool/CISettings.py -a IA32,X64 TOOL_CHAIN_TAG=GCC5
```

Hi. I have written down the steps I had to follow in order to run the tests. 

- I would like to clarify, do edk2 tests really need to be Rust dependent? I couldn't find any Rust code in the repository. 
- Is CSpell testing really necessary? It seems to be mostly ignored. Also would like to make this check optional if it is needed, as bringing dependencies in the form of nodejs is not wanted. 
- I would like to replace NuGet dependencies with native Linux tools.

I think I should be able to get some preliminary results in a couple of weeks. I am in the process of building CI for our Linux environment. I would like to keep the resources to run the tests low.
